### PR TITLE
[TC-IDM-10.1] Verify invalid/not allowed attribute and command id ranges

### PR DESCRIPTION
**Pull request by ReneJosefsen**
This is a follow up / complementery PR to https://github.com/project-chip/connectedhomeip/pull/28902.

This script goes through each available endpoint and cluster provided in a wildcard read and checks if any elements in the Attribute, Accepted Command or Generated Command Lists are out of the allowed range as stated by the latest test plan format for 1.1 test cases.

**For attributes:**
***
_The list SHALL NOT contain any values in the Test Vendor or invalid range: (0x0000_5000 - 0x0000_EFFF and 0x0000_FFFF), (0xXXXX_5000 - 0xXXXX_FFFF) and (0xFFF1_0000 - 0xFFFF_FFFF), where XXXX is the allowed MEI range (0x0001 - 0xFFF1)_
***

**For commands:**
***
_The list SHALL NOT contain any values in the Test Vendor or invalid range: (0x0000_0100 - 0x0000_FFFF), (0xXXXX_0100 - 0xXXXX_FFFF) and (0xFFF1_0000 - 0xFFFF_FFFF), where XXXX is the allowed MEI range (0x0001 - 0xFFF1)_
***

**And is prepared to check for Events:**
***
_The list SHALL NOT contain any values in the Test Vendor or invalid range: (0x0000_0100 - 0x0000_FFFF), (0xXXXX_0100 - 0xXXXX_FFFF) and (0xFFF1_0000 - 0xFFFF_FFFF), where XXXX is the allowed MEI range (0x0001 - 0xFFF1)_
***

Note that the above statement seems to have an issue where the allowed MEI range is listed as 0x0001 - 0xFFF1, but should be 0x0001 - 0xFFF0, per spec. The script does use 0x0001 - 0xFFF0 and this issue is being addressed in the test plan.

When 28902 merges and the test steps have been updated, this might as well be moved into the IDM-10.1 test and the test is written in a similar style to the remaining test of IDM-10.1, so it should not be an issue to move this later.

### CODE REVIEW

File: src/python_testing/TC_XX_1_1.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+                # print(f"EventList invalid range: {(standard_range_prefix + event_id_range_max + 1):#010x} - {(standard_range_prefix + event_id_invalid_max):#010x}")
+
+                if any((event_id > (standard_range_prefix + event_id_range_max) and event_id <= (standard_range_prefix + event_id_invalid_max)) for event_id in event_list):
+                    asserts.fail(f'Invalid event in range [{(standard_range_prefix + event_id_range_max + 1):#010x} - {(standard_range_prefix + event_id_invalid_max):#010x}]')
+                '''
+
+                # -- Commands
+                # print(f"CommandLists invalid range: {(standard_range_prefix + command_id_range_max + 1):#010x} - {(standard_range_prefix + command_id_invalid_max):#010x}")
+
+                if any((accepted_command_id > (standard_range_prefix + command_id_range_max) and accepted_command_id <= (standard_range_prefix + command_id_invalid_max)) for accepted_command_id in accepted_command_list):
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_XX_1_1.py:95]: This is very hard to parse.

I think this would be much clearer if we had:

1. Functions to extract the prefix/suffix.
2. Functions to check for suffix being in various valid ranges.
3. Functions to check for prefix being in various valid ranges.

Similar for the attributes cases above....  And given that we have to repeat everything for accepted/generated commands, ideally some sort of "valid command id" function that we can use for either....

Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py:95]: It is quite dense, I agree. 😄 

I will give it another spin and try to make it easier to read what it is actually doing as well as adding the suggested functions.

Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: Thanks for the great feedback, it made me rethink the approach and I think it got way simpler and easier to understand.

File: src/python_testing/TC_XX_1_1.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+# EVENT_LIST_ID = 0xFFFA
+ACCEPTED_COMMAND_LIST_ID = 0xFFF9
+GENERATED_COMMAND_LIST_ID = 0xFFF8
+
+
+def get_element_prefix_suffix(element):
+    prefix_mask = ~0xffff
+    suffix_mask = 0xffff
+
+    prefix = (element & prefix_mask) >> 16
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_XX_1_1.py:55]: How is this different from just:
```suggestion
    prefix = (element >> 16)
```
?

File: src/python_testing/TC_XX_1_1.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    elif prefix >= manufacturer_code_min_prefix and prefix <= manufacturer_code_max_prefix:
+        if suffix > attribute_id_range_max:
+            asserts.fail(f"Invalid attribute id {attribute_id:010x} in MC range")
+
+    elif prefix >= test_vendor_min_prefix and prefix <= test_vendor_max_prefix:
+        if suffix >= attribute_id_range_min:
+            asserts.fail(f"Invalid attribute id {attribute_id:010x} in Test Vendor MC range")
+
+    else:  # prefix >= invalid_vendor_min_prefix and prefix >= invalid_vendor_max_prefix
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_XX_1_1.py:76]:
```suggestion
    else:  # prefix >= invalid_vendor_min_prefix and prefix <= invalid_vendor_max_prefix
```

File: src/python_testing/TC_XX_1_1.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        if suffix >= attribute_id_range_min:
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_XX_1_1.py:77]: Shouldn't this fail unconditionally, since this is not a valid MEI prefix?

File: src/python_testing/TC_XX_1_1.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    if prefix == standard_range_prefix:
+        if suffix > attribute_id_range_max and suffix < attribute_global_range_min or suffix == attribute_id_invalid_max:
+            asserts.fail(f"Invalid attribute id {attribute_id:010x} in standard range")
+
+    elif prefix >= manufacturer_code_min_prefix and prefix <= manufacturer_code_max_prefix:
+        if suffix > attribute_id_range_max:
+            asserts.fail(f"Invalid attribute id {attribute_id:010x} in MC range")
+
+    elif prefix >= test_vendor_min_prefix and prefix <= test_vendor_max_prefix:
+        if suffix >= attribute_id_range_min:
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_XX_1_1.py:73]: This looks like it fails for all attribute ids, but in that case, why are we checking the suffix at all?  This at least needs a comment describing the intended behavior...

File: src/python_testing/TC_XX_1_1.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    if prefix == standard_range_prefix:
+        if suffix > command_id_range_max:
+            asserts.fail(f"Invalid command id {command_id:010x} in standard range")
+
+    elif prefix >= manufacturer_code_min_prefix and prefix <= manufacturer_code_max_prefix:
+        if suffix > command_id_range_max:
+            asserts.fail(f"Invalid command id {command_id:010x} in MC range")
+
+    elif prefix >= test_vendor_min_prefix and prefix <= test_vendor_max_prefix:
+        if suffix >= command_id_range_min:
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_XX_1_1.py:93]: Again, this is going to fail for all command ids, but it's obfuscating that somewhat...

File: src/python_testing/TC_XX_1_1.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            asserts.fail(f"Invalid command id {command_id:010x} in Test Vendor MC range")
+
+    else:  # prefix >= invalid_vendor_min_prefix and prefix >= invalid_vendor_max_prefix
+        if suffix >= command_id_range_min:
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_XX_1_1.py:97]: This should defintely just fail unconditionally, due to not being a valid MEI.

**Review by bzbarsky-apple — COMMENTED** [review]
This is much clearer!

File: src/python_testing/TC_XX_1_1.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+invalid_vendor_min_prefix = 0xFFF5
+invalid_vendor_max_prefix = 0xFFFF
+
+attribute_id_range_min = 0x0000
+attribute_id_range_max = 0x4FFF
+attribute_global_range_min = 0xF000
+attribute_global_range_max = 0xFFFE
+attribute_id_invalid_max = 0xFFFF
+
+# event_id_range_min = 0x00
```
Comment (andy31415) [inline src/python_testing/TC_XX_1_1.py:37]: remove commented out code.
For constants I would prefer `UPPER_SNAKE_CASE` throughout (it seems IDs are like that but min/max are not.

Instead of having separate constants for min/max, how about merging them into one thing. Supposedly range allows for fixed sizeso you can do:

```py
MANUFACTURER_CODE_RANGE = range(0x0001, 0xFFF0 + 1)
TEST_VENDOR_RANGE = range(0xFFF1, 0xFFF4 + 1)
# ...
```

and then check range membership

File: src/python_testing/TC_XX_1_1.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+command_id_range_max = 0xFF
+command_id_invalid_max = 0xFFFF
+
+ATTRIBUTE_LIST_ID = 0xFFFB
+# EVENT_LIST_ID = 0xFFFA
+ACCEPTED_COMMAND_LIST_ID = 0xFFF9
+GENERATED_COMMAND_LIST_ID = 0xFFF8
+
+
+def get_element_prefix_suffix(element):
```
Comment (andy31415) [inline src/python_testing/TC_XX_1_1.py:51]: is `split_element_into_prefix_suffix` a clearer name?

File: src/python_testing/TC_XX_1_1.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    suffix = (element & suffix_mask)
+
+    return prefix, suffix
+
+
+def is_attribute_id_in_valid_range(attribute_id):
+    prefix, suffix = get_element_prefix_suffix(attribute_id)
+
+    if prefix == standard_range_prefix:
+        if suffix > attribute_id_range_max and suffix < attribute_global_range_min or suffix == attribute_id_invalid_max:
```
Comment (andy31415) [inline src/python_testing/TC_XX_1_1.py:65]: what rule does this enforce?
it seems easier to read if we would have:

```py
if suffix not in ATTRIBUTE_ID_RANGE and suffix not in GLOBAL_ATTRIBUTE_ID_RANGE:
   asserts.fail(...)
```

File: src/python_testing/TC_XX_1_1.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+                logging.info(f"Verifying ClusterID: {cluster_id:#06x}")
+
+                attribute_list = cluster[ATTRIBUTE_LIST_ID]
+                accepted_command_list = cluster[ACCEPTED_COMMAND_LIST_ID]
+                generated_command_list = cluster[GENERATED_COMMAND_LIST_ID]
+
+                # -- Attributes
+                logging.info(f"AttributeList: {attribute_list}")
+                for attribute_id in attribute_list:
+                    is_attribute_id_in_valid_range(attribute_id)
```
Comment (andy31415) [inline src/python_testing/TC_XX_1_1.py:125]: is_* sounds like a method returning a bool. Should maybe be named `validate_attribute_id_range` or similar

File: src/python_testing/TC_XX_1_1.py (hunk 2 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py:55]: Since it is bitshifting anyways, you are right. The mask dones not really do anything here.

File: src/python_testing/TC_XX_1_1.py (hunk 5 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py:73]: Yeah, I get your point. If the complete range is disallowed, there is no need to hide it being what looks like a range check. I will update this.

Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py:73]: For the test vendor ID, it was discussed to not outright fail since the CI/SDK uses this for some testing elements, so I will try to update the script to be inline with this.

The test plan should also have been updated to change this section to a SHOULD NOT statement.

File: src/python_testing/TC_XX_1_1.py (hunk 8 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py:37]: I have no problem with this change, I mainly tried to match the way https://github.com/project-chip/connectedhomeip/pull/28902/files is written.

File: src/python_testing/TC_XX_1_1.py (hunk 9 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py:51]: Fine with me, I will change it.

File: src/python_testing/TC_XX_1_1.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+
+def get_element_prefix_suffix(element):
+    prefix_mask = ~0xffff
+    suffix_mask = 0xffff
+
+    prefix = (element & prefix_mask) >> 16
+    suffix = (element & suffix_mask)
+
+    return prefix, suffix
```
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py:58]: I will clean it up 👍🏻

File: src/python_testing/TC_XX_1_1.py (hunk 11 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py:125]: Agree, the "is" word does give that impression. So I will update it.

File: src/python_testing/TC_XX_1_1.py (hunk 4 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: This is addressed in the latest commit

File: src/python_testing/TC_XX_1_1.py (hunk 5 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: This is addressed in the latest commit

File: src/python_testing/TC_XX_1_1.py (hunk 6 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: This is addressed in the latest commit

File: src/python_testing/TC_XX_1_1.py (hunk 7 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: This is addressed in the latest commit

File: src/python_testing/TC_XX_1_1.py (hunk 8 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: I have tried reworking the test script around your suggestion, let me know if this is acceptable.

Maybe we should consider doing a similar approach in IDM-10.1 once we merge these two test scripts.

File: src/python_testing/TC_XX_1_1.py (hunk 9 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: This is addressed in the latest commit

File: src/python_testing/TC_XX_1_1.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: This is addressed in the latest commit

File: src/python_testing/TC_XX_1_1.py (hunk 11 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: This is addressed in the latest commit

File: src/python_testing/TC_XX_1_1.py (hunk 10 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_XX_1_1.py]: This verifies that a DUT does not have any attributes outside of the allowed attribute range and global attribute range as defined by spec in https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Data-Model-Namespace.adoc#ref_mei_data_model_types

I have updated the checks based on your suggestions, let me know if this is acceptable.

File: src/python_testing/TC_IDM_10_1.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+from matter_testing_support import MatterBaseTest, async_test_body, default_matter_test_main
+from mobly import asserts
+
+STANDARD_RANGE = 0x0000
+MANUFACTURER_CODE_RANGE = range(0x0001, 0xFFF0 + 1)
+TEST_VENDOR_RANGE = range(0xFFF1, 0xFFF4 + 1)
+INVALID_VENDOR_RANGE = range(0xFFF5, 0xFFFF + 1)
+
+ATTRIBUTE_ID_RANGE = range(0x0000, 0x4FFF + 1)
+GLOBAL_ATTRIBUTE_ID_RANGE = range(0xF000, 0xFFFE)
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_IDM_10_1.py:29]:
```suggestion
GLOBAL_ATTRIBUTE_ID_RANGE = range(0xF000, 0xFFFE + 1)
```

File: src/python_testing/TC_IDM_10_1.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+def split_element_into_prefix_suffix(element):
+    return (element >> 16), (element & 0xFFFF)
+
+
+def validate_attribute_id_range(attribute_id):
+    prefix, suffix = split_element_into_prefix_suffix(attribute_id)
+
+    if prefix == STANDARD_RANGE:
+        if suffix not in ATTRIBUTE_ID_RANGE and suffix not in GLOBAL_ATTRIBUTE_ID_RANGE or suffix == ATTRIBUTE_INVALID_MAX:
```
Comment (bzbarsky-apple) [inline src/python_testing/TC_IDM_10_1.py:48]:
```suggestion
        if suffix not in ATTRIBUTE_ID_RANGE and suffix not in GLOBAL_ATTRIBUTE_ID_RANGE :
```
because being not in both those ranges would include the max value....

**Review by bzbarsky-apple — APPROVED** [review]

File: src/python_testing/TC_IDM_10_1.py (hunk 14 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_IDM_10_1.py:48]: Oh yes, true. This seems to have been left over (in my mind) from when the check was reversed 😄 Thanks,

File: src/python_testing/TC_IDM_10_1.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+
+STANDARD_RANGE = 0x0000
+MANUFACTURER_CODE_RANGE = range(0x0001, 0xFFF0 + 1)
+TEST_VENDOR_RANGE = range(0xFFF1, 0xFFF4 + 1)
+INVALID_VENDOR_RANGE = range(0xFFF5, 0xFFFF + 1)
+
+ATTRIBUTE_ID_RANGE = range(0x0000, 0x4FFF + 1)
+GLOBAL_ATTRIBUTE_ID_RANGE = range(0xF000, 0xFFFE)
+ATTRIBUTE_INVALID_MAX = 0xFFFF
+
```
Comment (ReneJosefsen) [inline src/python_testing/TC_IDM_10_1.py:31]:
```suggestion

```

File: src/python_testing/TC_IDM_10_1.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+STANDARD_RANGE = 0x0000
+MANUFACTURER_CODE_RANGE = range(0x0001, 0xFFF0 + 1)
+TEST_VENDOR_RANGE = range(0xFFF1, 0xFFF4 + 1)
+INVALID_VENDOR_RANGE = range(0xFFF5, 0xFFFF + 1)
+
+ATTRIBUTE_ID_RANGE = range(0x0000, 0x4FFF + 1)
+GLOBAL_ATTRIBUTE_ID_RANGE = range(0xF000, 0xFFFE + 1)
+COMMAND_ID_RANGE = range(0x00, 0xFF + 1)
+COMMAND_ID_INVALID_MAX = 0xFFFF
```
Comment (ReneJosefsen) [inline src/python_testing/TC_IDM_10_1.py:31]:
```suggestion
```

**Review by tima-q — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
